scripts/release.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+from typing import Generator, List, Optional, Tuple
+import argparse
+import json
+import os
+import re
+import subprocess
+import sys
+
+# Internal
+PACKAGE_PATH = os.path.realpath(os.path.join(os.path.join(os.path.dirname(__file__), '..')))
+MESSAGE_DIR = 'messages'
+MESSAGE_PATH = os.path.join(PACKAGE_PATH, MESSAGE_DIR)
+
+with open(os.path.join(PACKAGE_PATH, '.release.json'), 'r') as f:
+    CONFIGURATION = json.load(f)
+
+# Project configuration
+# The name of the branch to push to the remote on releasing.
+RELEASE_BRANCH = CONFIGURATION['push_branch']
+# The name of the GitHub repository in <owner>/<repo> format
+GITHUB_REPO = CONFIGURATION['publish_repo']
+# The name of the settings file to get the release token from ("github_token" setting)
+SETTINGS = '{}.sublime-settings'.format(__package__)
+PYTHON_VERSION_PATH = CONFIGURATION.get('python_version_path', None)
+
+
+def get_message(fname: str) -> str:
+    with open(fname, 'r', encoding='utf-8') as file:
+        message = file.read()
+    return message
+
+
+def put_message(fname: str, text: str) -> None:
+    with open(fname, 'w', encoding='utf-8') as file:
+        file.write(text)
+
+
+def build_messages_json(version_history: List[str]) -> None:
+    """Write the version history to the messages.json file."""
+    output = os.path.join(PACKAGE_PATH, 'messages.json')
+    with open(output, 'w+', encoding='utf-8') as file:
+        json.dump(
+            obj={v: MESSAGE_DIR + '/' + v + '.txt' for v in version_history},
+            fp=file, indent=4, separators=(',', ': '), sort_keys=True)
+        file.write('\n')
+
+
+def version_history() -> List[str]:
+    """Return a list of all releases."""
+    def generator() -> Generator[str, None, None]:
+        for filename in os.listdir(MESSAGE_PATH):
+            basename, ext = os.path.splitext(filename)
+            if ext.lower() == '.txt':
+                yield basename
+
+    return sorted(tuple(generator()), key=parse_version)
+
+
+def parse_version(version: str) -> Tuple[int, int, int]:
+    """Convert filename to version tuple (major, minor, patch)."""
+    match = re.match(
+        r'(?:(?P<prefix>[^.-]+)\-)?(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)(?:-.+)?', version)
```

Comment:
LSP-* plugins don't follow the same version strategy as LSP or any version (concrete or agreed) strategy really. 

If you see this is how the LSP-typescript is setup on package control.
https://github.com/sublimelsp/package_control_channel/commit/23805a24b5273476e30bf385fdef923f3a2e4618#diff-5108129fa257c68bb90628d08872bdddR1736-R1745
```
{
			"name": "LSP-typescript",
			"details": "https://github.com/sublimelsp/LSP-typescript",
			"releases": [
				{
					"sublime_text": ">=3092",
					"tags": true
				}
			]
		},
``` 
and this is LSP:
```
"releases": [
				{
					"sublime_text": "3154 - 4069",
					"tags": "3154-"
				},
				{
					"sublime_text": ">=4070",
					"tags": "4070-"
				}
			]
```

Reply:
Ignore this, I was just reading code line by line.
I didn't understand all the steps that are required to use the script.

So to use the script.
I only need to make sure to create a file in `messages/(OPTIONAL_PREFIX)?-MAJOR.MINOR.PATCH.txt
and run the script? 
The script will base on that name, update the VERSIONS file, and publish a release at github

Reply:
Yes, you need to create a message file to release right now but that's why this is in a draft state - I don't want to always have to create a message file when releasing a new version as I don't think user should always be bothered with a changelog if it's not significant.

I want to have an option to release without creating the message file. So I have plans to add an optional feature to specify a new version (patch, minor, major) on the command line, skipping the message file. Although in that case, we would have no release notes for github release so it's not ideal either.

The way release scripts handle such things is for example:
 a) to have a global CHANGELOG.md file where I'd log my changes and those would be picked up on release
 b) get changes from git log

Getting changes from git log is not ideal because often we want to skip some non-important commits in release notes.
For handling global changelog I'd need to write quite some code to parse it and extract into messages/* files when needed.

Just thinking about it gets a bit overwhelming and I don't feel motivated enough to do that right now as it's not that important. If using NPM, I'd just use some existing script for that (although handling messages/* would still need to be done separately).